File: Алгоритмы/Лаб.4/MergeSort.py
import time


def mergeSortTimer(array):
    start = time.perf_counter()
    def mergeSort(array):
        if len(array) == 1:
            return array
        mid = len(array) // 2
        # Делим на 2 части правлую и левую
        right = mergeSort(array[mid:]) # центральный елемент включительно
        left = mergeSort(array[:mid])

        return merge(left, right)

    def merge(left, right):
        merged = [0 for x in range(len(left)+len(right))] # пустой список
        left_array_index = 0
        right_array_index = 0
        k = 0
        # Сортировка +добовление элементов в масив
        while left_array_index < len(left) and right_array_index < len(right):
            if left[left_array_index] < right[right_array_index]:
                merged[k] = left[left_array_index]
                left_array_index += 1
            else:
                merged[k] = right[right_array_index]
                right_array_index += 1
            k += 1
        while right_array_index < len(right):
            merged[k] = right[right_array_index]
            right_array_index += 1
            k += 1
        while left_array_index < len(left):
            merged[k] = left[left_array_index]
            left_array_index += 1
            k += 1

        return merged

    mergeSort(array)
    end = time.perf_counter()
    finalTime = 1000 * (end - start)
    return "Merge sort for "+ str(len(array)) + " elements took "+ str(round(finalTime,3))+" miliseconds"

File: Year 1/Semester 1/Basics of Programming/Lab 7/cabbage/MergeSort.java
package cabbage;

public class MergeSort implements SortChooser {
	void merge(float arr[], int l, int m, int r){
		int n1 = m - l + 1;
		int n2 = r - m;
		float L[] = new float[n1];
		float R[] = new float[n2];
		for (int i = 0; i < n1; ++i)
			L[i] = arr[l + i];
		for (int j = 0; j < n2; ++j)
			R[j] = arr[m + 1 + j];
		int i = 0, j = 0;
		int k = l;
		while (i < n1 && j < n2) {
			if (L[i] <= R[j]) {
				arr[k] = L[i];
				i++;
			}
			else {
				arr[k] = R[j];
				j++;
			}
			k++;
		}
		while (i < n1) {
			arr[k] = L[i];
			i++;
			k++;
		}
		while (j < n2) {
			arr[k] = R[j];
			j++;
			k++;
		}
	}
	void sorter(float arr[], int l, int r){
		if (l < r) {
			int m = (l + r) / 2;
			sorter(arr, l, m);
			sorter(arr, m + 1, r);
			merge(arr, l, m, r);
		}
	}
	public void sort (float[] arr) {
		sorter(arr, 0, arr.length-1);
	}
}